src/charm.py

```diff
@@ -219,27 +235,74 @@ def self_scraping_job(self):
         """The self-monitoring scrape job."""
         external_url = urlparse(self._external_url)
         metrics_path = f"{external_url.path.rstrip('/')}/metrics"
-        target = (
-            f"{external_url.hostname}{':'+str(external_url.port) if external_url.port else ''}"
-        )
+        internal_url = self._internal_url.replace("http://", "")
+        target = f"{internal_url}"
         job = {
             "metrics_path": metrics_path,
             "static_configs": [{"targets": [target]}],
         }
 
         return [job]
 
+    def _update_blackbox_config_yaml_from_relation(self, modules) -> None:
+        """Update the blackbox config yaml with modules defined in relation.
+
+        This function takes the modules from the BlackboxExporterRequirer and
+        updates the config file with the required modules.
+        """
+        config_file_data = self.container.pull(self._config_path).read()
+
+        if not config_file_data:
+            return
+
+        config_data = yaml.safe_load(config_file_data)
```

Comment:
- Please add to docstring a "Raises:" section for Pebble and yaml exceptions.
- Do we have a unit test for config file with invalid yaml?